### ISSUE TITLE
Compile error with rustc 1.16.0-nightly (4ecc85beb 2016-12-28)

### ISSUE THREAD

**Issue by choleraehyq** [Manishearth/rust-clippy#1412]
```
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling quine-mc_cluskey v0.2.4
   Compiling unicode-normalization v0.1.3
   Compiling regex-syntax v0.3.9
   Compiling matches v0.1.4
   Compiling rustc-serialize v0.3.22
   Compiling nom v1.2.4
   Compiling semver v0.2.3
   Compiling toml v0.1.30
   Compiling clippy_lints v0.0.104
error[E0432]: unresolved import `rustc::middle::const_qualif::ConstQualif`
 --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/methods.rs:4:5
  |
4 | use rustc::middle::const_qualif::ConstQualif;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `const_qualif` in `middle`

error[E0531]: unresolved tuple struct/variant `MethodTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/attrs.rs:171:9
    |
171 |         MethodTraitItem(_, None) => true,
    |         ^^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `MethodTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/attrs.rs:172:9
    |
172 |         MethodTraitItem(_, Some(eid)) => is_relevant_expr(cx, cx.tcx.map.expr(eid)),
    |         ^^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `MethodTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/cyclomatic_complexity.rs:109:16
    |
109 |         if let MethodTraitItem(_, Some(eid)) = item.node {
    |                ^^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `hir::MethodTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/functions.rs:109:16
    |
109 |         if let hir::MethodTraitItem(ref sig, eid) = item.node {
    |                ^^^^^^^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `MethodTraitItem`
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/len_zero.rs:93:16
   |
93 |         if let MethodTraitItem(ref sig, _) = item.node {
   |                ^^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `MethodTraitItem`
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/lifetimes.rs:73:16   |
73 |         if let MethodTraitItem(ref sig, _) = item.node {
   |                ^^^^^^^^^^^^^^^ no resolution found

error[E0412]: unresolved type `hir::ExplicitSelf`
    --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/methods.rs:1337:28
     |
1337 |     fn matches(self, slf: &hir::ExplicitSelf, allow_value_for_ref: bool) -> bool {
     |                            ^^^^^^^^^^^^^^^^^ no resolution found
     |
     = help: possible candidate is found in another module, you can import it into scope:
     = help:   `use syntax::ast::ExplicitSelf;`

error[E0433]: failed to resolve. Could not find `SelfKind` in `hir`
    --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/methods.rs:1339:32
     |
1339 |             (SelfKind::Value, &hir::SelfKind::Value(_)) |
     |                                ^^^^^^^^^^^^^^^^^^^^ Could not find `SelfKind` in `hir`

error[E0433]: failed to resolve. Could not find `SelfKind` in `hir`
    --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/methods.rs:1340:30
     |
1340 |             (SelfKind::Ref, &hir::SelfKind::Region(_, hir::Mutability::MutImmutable)) |
     |                              ^^^^^^^^^^^^^^^^^^^^^ Could not find `SelfKind` in `hir`

error[E0433]: failed to resolve. Could not find `SelfKind` in `hir`
    --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/methods.rs:1341:33
     |
1341 |             (SelfKind::RefMut, &hir::SelfKind::Region(_, hir::Mutability::MutMutable)) => true,
     |                                 ^^^^^^^^^^^^^^^^^^^^^ Could not find `SelfKind` in `hir`

error[E0433]: failed to resolve. Could not find `SelfKind` in `hir`
    --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/methods.rs:1342:30
     |
1342 |             (SelfKind::Ref, &hir::SelfKind::Value(_)) |
     |                              ^^^^^^^^^^^^^^^^^^^^ Could not find `SelfKind` in `hir`

error[E0433]: failed to resolve. Could not find `SelfKind` in `hir`
    --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/methods.rs:1343:33
     |
1343 |             (SelfKind::RefMut, &hir::SelfKind::Value(_)) => allow_value_for_ref,
     |                                 ^^^^^^^^^^^^^^^^^^^^ Could not find `SelfKind` in `hir`

error[E0433]: failed to resolve. Could not find `SelfKind` in `hir`
    --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/methods.rs:1344:18
     |
1344 |             (_, &hir::SelfKind::Explicit(ref ty, _)) => self.matches_explicit_type(ty, allow_value_for_ref),
     |                  ^^^^^^^^^^^^^^^^^^^^^^^ Could not find `SelfKind` in `hir`

error[E0531]: unresolved tuple struct/variant `hir::ConstTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/missing_doc.rs:139:13
    |
139 |             hir::ConstTraitItem(..) => "an associated constant",
    |             ^^^^^^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `hir::MethodTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/missing_doc.rs:140:13
    |
140 |             hir::MethodTraitItem(..) => "a trait method",
    |             ^^^^^^^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `hir::TypeTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/missing_doc.rs:141:13
    |
141 |             hir::TypeTraitItem(..) => "an associated type",
    |             ^^^^^^^^^^^^^^^^^^ no resolution found

error[E0425]: unresolved function `print::path_to_string`
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/ok_if_let.rs:55:16   |
45 |         if_let_chain! {[ //begin checking variables
   |         - in this macro invocation
...
55 |             if print::path_to_string(x) == "Some" && is_result_type {
   |                ^^^^^^^^^^^^^^^^^^^^^ no resolution found
   |
   = help: possible candidate is found in another module, you can import it into scope:
   = help:   `use syntax::print::pprust::path_to_string;`

error[E0531]: unresolved tuple struct/variant `MethodTraitItem`
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/ptr.rs:76:16
   |
76 |         if let MethodTraitItem(ref sig, _) = item.node {
   |                ^^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `ConstTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/types.rs:559:13
    |
559 |             ConstTraitItem(ref ty, _) |
    |             ^^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `TypeTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/types.rs:560:13
    |
560 |             TypeTraitItem(_, Some(ref ty)) => self.check_type(cx, ty),
    |             ^^^^^^^^^^^^^ no resolution found

error[E0531]: unresolved tuple struct/variant `MethodTraitItem`
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/types.rs:561:13
    |
561 |             MethodTraitItem(MethodSig { ref decl, .. }, None) => self.check_fndecl(cx, decl),
    |             ^^^^^^^^^^^^^^^ no resolution found

error[E0053]: method `check_fn` has an incompatible type for trait
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/escape.rs:64:5
   |
64 |     fn check_fn(
   |     ^ expected struct `rustc::hir::Body`, found struct `rustc::hir::Expr`
   |
   = note: expected type `fn(&mut escape::Pass, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Body, syntax::codemap::Span, syntax::ast::NodeId)`
   = note:    found type `fn(&mut escape::Pass, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Expr, syntax::codemap::Span, syntax::ast::NodeId)`

error[E0053]: method `check_fn` has an incompatible type for trait
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/functions.rs:72:5    |
72  |     fn check_fn(
    |     ^ expected struct `rustc::hir::Body`, found struct `rustc::hir::Expr`
    |
    = note: expected type `fn(&mut functions::Functions, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Body, syntax::codemap::Span, syntax::ast::NodeId)`
    = note:    found type `fn(&mut functions::Functions, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Expr, syntax::codemap::Span, syntax::ast::NodeId)`

error[E0053]: method `check_fn` has an incompatible type for trait
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/misc.rs:170:5
    |
170 |     fn check_fn(&mut self, cx: &LateContext<'a, 'tcx>, k: FnKind<'tcx>, decl: &'tcx FnDecl, _: &'tcx Expr, _: Span, _: NodeId) {
    |     ^ expected struct `rustc::hir::Body`, found struct `rustc::hir::Expr`
    |
    = note: expected type `fn(&mut misc::Pass, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Body, syntax::codemap::Span, syntax::ast::NodeId)`
    = note:    found type `fn(&mut misc::Pass, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Expr, syntax::codemap::Span, syntax::ast::NodeId)`

error[E0053]: method `check_fn` has an incompatible type for trait
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/new_without_default.rs:93:5
    |
93  |     fn check_fn(
    |     ^ expected struct `rustc::hir::Body`, found struct `rustc::hir::Expr`
    |
    = note: expected type `fn(&mut new_without_default::NewWithoutDefault, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Body, syntax::codemap::Span, syntax::ast::NodeId)`
    = note:    found type `fn(&mut new_without_default::NewWithoutDefault, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Expr, syntax::codemap::Span, syntax::ast::NodeId)`

error[E0053]: method `check_fn` has an incompatible type for trait
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/shadow.rs:83:5
   |
83 |     fn check_fn(
   |     ^ expected struct `rustc::hir::Body`, found struct `rustc::hir::Expr`
   |
   = note: expected type `fn(&mut shadow::Pass, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Body, syntax::codemap::Span, syntax::ast::NodeId)`
   = note:    found type `fn(&mut shadow::Pass, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Expr, syntax::codemap::Span, syntax::ast::NodeId)`

error[E0053]: method `check_fn` has an incompatible type for trait
   --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/types.rs:539:5
    |
539 |       fn check_fn(&mut self, cx: &LateContext<'a, 'tcx>, _: FnKind<'tcx>, decl: &'tcx FnDecl, _: &'tcx Expr, _: Span, _: NodeId) {
    |  _____^ starting here...
540 | |         self.check_fndecl(cx, decl);
541 | |     }
    | |_____^ ...ending here: expected struct `rustc::hir::Body`, found struct `rustc::hir::Expr`    |
    = note: expected type `fn(&mut types::TypeComplexityPass, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Body, syntax::codemap::Span, syntax::ast::NodeId)`
    = note:    found type `fn(&mut types::TypeComplexityPass, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Expr, syntax::codemap::Span, syntax::ast::NodeId)`

error[E0053]: method `check_fn` has an incompatible type for trait
  --> .cargo/registry/src/github.com-1ecc6299db9ec823/clippy_lints-0.0.104/src/unused_label.rs:44:5
   |
44 |     fn check_fn(
   |     ^ expected struct `rustc::hir::Body`, found struct `rustc::hir::Expr`
   |
   = note: expected type `fn(&mut unused_label::UnusedLabel, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Body, syntax::codemap::Span, syntax::ast::NodeId)`
   = note:    found type `fn(&mut unused_label::UnusedLabel, &rustc::lint::LateContext<'a, 'tcx>, rustc::hir::intravisit::FnKind<'tcx>, &'tcx rustc::hir::FnDecl, &'tcx rustc::hir::Expr, syntax::codemap::Span, syntax::ast::NodeId)`

error: aborting due to 7 previous errors

error: failed to compile `clippy v0.0.104`, intermediate artifacts can be found at `/tmp/cargo-install.eK0L6eOPrpdW`

Caused by:
  Could not compile `clippy_lints`.

To learn more, run the command again with --verbose.
```

**Comment by Maplicant** [comment]
Having the same exact problem. It might be relevant that I'm using Windows.

**Comment by choleraehyq** [comment]
@Maplicant 
I've met this error both on windows and fedora.

**Comment by shssoichiro** [comment]
Also present on OS X

**Comment by Manishearth** [comment]
We need to update to the latest nightly, platform independent. Will happen soon, sit tight.

**Comment by troplin** [comment]
I wonder why the Travis badge still says "build passing" and when you click on it you see the log of a failed build...

**Comment by llogiq** [comment]
"There are two hard things in computing: Cache invalidation, naming things and off-by-one errors.

**Comment by troplin** [comment]
Ah no, it's not a cache issue.
The last commit two days ago was still tested with nightly 1.15 and the current failing build is merely a PR, not a commit.
It's just a bit confusing that this is what is shown first when clicking on the badge.

**Comment by llogiq** [comment]
Who's working on this?

**Comment by Manishearth** [comment]
I'll have time to do this later today if nobody else gets to it.

**Comment by Manishearth** [comment]
Got some work done at https://github.com/Manishearth/rust-clippy/tree/rustup

The removal of explicitself complicates things.

**Comment by shssoichiro** [comment]
I'm poking around on @Manishearth's branch, but he may be able to solve this faster, so I won't stop him from pushing a fix if he's able to get to it first.

**Comment by Manishearth** [comment]
Not going to pick it up again till tomorrow, so go ahead. If you want help I hang out in #rust-internals and #rust-tools .

**Comment by shssoichiro** [comment]
I'm making progress on many of the simpler errors (I fixed 8 and another 67 popped up!), but I might need help figuring out the `SelfKind` issue in `clippy_lints/src/methods.rs`.

**Comment by Manishearth** [comment]
You'll need to check with `has_self()` first, get the type of the argument and of the impl, and pass to https://github.com/rust-lang/rust/blob/4ecc85beb339aa8089d936e450b0d800bdf580ae/src/librustc_typeck/astconv.rs#L2154

There are invocations of has_self in https://github.com/rust-lang/rust/commit/6ebb6fdbee7f0d250e50e4d84d9aede293cf3bb3 to compare with

**Comment by shssoichiro** [comment]
`has_self` was also removed in 2016-12-28, although I've copied the previous implementation over in a couple of other places where it broke clippy. I'll see what I can do with that in the morning. For now, I've committed what I have on [a branch here](https://github.com/shssoichiro/rust-clippy/tree/nightly-2016-12-28).

**Comment by shssoichiro** [comment]
I think I've gotten as far as I reasonably can on `clippy_lints`. I've pushed to the branch I linked in my previous comment. The remaining errors are mostly from enums that previously returned a `hir::Pat` or `hir::Arg` now returning a `hir::Ty` or `hir::NodeId`.

**Comment by shssoichiro** [comment]
The 2017-01-03 nightly broke a few more things, but these were easily fixable (s/u64/u128/). I've pushed a commit that fixes these as well.

**Comment by Manishearth** [comment]
Just a note, in Rust, variables are named like `body_id`, not `bodyId`. I've got it compiling but your fix for the self type is wrong (it should check the type, not the pattern). No biggie though, it mostly changes how edge cases are handled.

**Comment by shssoichiro** [comment]
Whoops, I knew that. Guess that's what happens when I use camel-case all day at work. Thanks for helping me through this one!

**Comment by Manishearth** [comment]
Finished rustup at https://github.com/Manishearth/rust-clippy/pull/1419 (will merge when tests pass)

Thanks for the help!

**Comment by Manishearth** [comment]
Published 0.0.105
